extmod/modbluetooth: Add pairing/bonding support.

jimmo opened the pull request:
WIP, looking for feedback on the approach.

Much of this work was done by @andrewleech in #6289, #6312, and #6447. This PR changes:
 - Moved the bonding secret store into Python (via _IRQ_GET_SECRET and _IRQ_SET_SECRET events).
 - Adds characteristic flags to enable auto-pairing (e.g. READ_ENCRYPTED)
 - Adds passkey authentication.
 - gap_pair just takes conn_handle, and pairing/bonding parameters are set via config().
 - Documentation

Currently only supported on NimBLE on STM32 and Unix, as it requires synchronous (non-ringbuffer) events. Enabled via `MICROPY_PY_BLUETOOTH_ENABLE_PAIRING_BONDING` macro (which is now enabled by default for NimBLE STM32/Unix). Some BlueKitchen support is implemented, but will need to wait until it's migrated to synchronous events. The secret store is designed to be compatible with BlueKitchen's TLV store API.

Tested against Android and BlueZ.

This PR adds:
 - Better error handling for oversize characteristic reads.
 - An _IRQ_CONNECTION_UPDATE event. Not specifically related to pairing and bonding, but useful information about the connection state.
 - An _IRQ_ENCRYPTION_UPDATE event to get notified about pairing/bonding state change.
 - Widens characteristic flags to 16-bit and allows for setting encryption/authentication/authorisation requirements.
 - Allows the user to fail the read request IRQ with a custom error code.
 - Configuration of pairing/bonding parameters (le_secure, bond, mitm, io)
 - gap_pair() method to initiate pairing.
 - _IRQ_GET_SECRET and IRQ_SET_SECRET events for persistent key management.
 - Simplify the handing of numeric arguments in invoke_irq_handler (used in synchronous events).
 - An _IRQ_PASSKEY_ACTION event to be notified of an authentication request.
 - gap_passkey() method to respond to an authentication request.
 - ble_bonding_peripheral.py example, showing implementation of key persistence and authentication. (WIP)

Four major outstanding pieces:
 - Write multi-test.
 - Support address resolving in central mode (perhaps via a _IRQ_RESOLVED_SCAN_RESULT event)
 - Improve handing of RPA mode. NimBLE doesn't seem to support using a non-fixed IRK (see https://github.com/apache/mynewt-nimble/issues/887)
 - PYBD doesn't work with RPA mode. (Works fine on Unix+Zephyr and WB55)

pi-anl committed:
extmod/modbluetooth: Allow setting char/desc enc/auth options.

This widens the characteristic/descriptor flags to 16-bit, to allow setting
encryption/authentication requirements.

Sets the required flags for NimBLE and btstack implementations.

Deprecate the BLE.FLAG_* constants in favour of copy & paste Python constants
(like the IRQs).

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

pi-anl committed:
extmod/modbluetooth: Add ENCRYPTION_UPDATE event.

This allows the application to be notified if any of
{encrypted,authenticated,bonded} states change as well as the encryption key
size.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

pi-anl committed:
extmod/modbluetooth: Add CONNECTION_UPDATE event.

This allows the application to be notified of changes to
 - Connection interval
 - Connection latency
 - Supervision timeout

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
docs/library/ubluetooth.rst: Update char/desc flags.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/modbluetooth: Allow user-specified reason in read requst IRQ.

Instead of returning None/bool from the IRQ, return None/int.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
docs/library/ubluetooth.rst: Update read request IRQ.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
docs/library/ubluetooth.rst: Add _IRQ_CONNECTION_UDPATE docs.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/modbluetooth: Allow configuration of pair/bond parameters.

Sets security and MITM-protection requirements.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
docs/library/ubluetooth.rst: Add pairing/bonding config docs.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/modbluetooth: Add gap_pair(conn_handle) to intiate pairing.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
docs/library/ubluetooth.rst: Add gap_pair() docs.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/nimble: Fail read if the characteristic is too big.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/modbluetooth: Add support for bonding (key persistence).

This adds `_IRQ_GET_SECRET` and `_IRQ_SET_SECRET` to allow
the stack to request the Python code retrive/store/delete
secret key data.

The actual keys and values are opaque to Python and stack-specific.

Only NimBLE is implemented (pending moving btstack to sync events).

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/modbluetooth: Add flag to enable pairing/bonding.

Enable for STM32/Unix NimBLE only (requires synchronous events and full bindings).

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/modbluetooth.c: Simplify sync invoke_irq_handler.

Rather than dealing with the different int types, just pass them all as a single array of mp_int_t with n_unsigned (before addr) and n_signed (after addr).

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/modbluetooth: Add support for passkey auth.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
examples/bluetooth: Add bonding/passkey demo.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
docs/library/ubluetooth.rst: Add bonding docs.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
docs/library/ubluetooth.rst: Add passkey docs.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

dmazzella commented:
+1 for bonding secret store into Python

andrewleech commented:
Minor note: in top post `MICROPY_BLUETOOTH_ENABLE_PAIRING_BONDING` should be `MICROPY_PY_BLUETOOTH_ENABLE_PAIRING_BONDING`

andrewleech commented on extmod/nimble/modbluetooth_nimble.c:
```diff
@@ -808,6 +889,40 @@ int mp_bluetooth_set_preferred_mtu(uint16_t mtu) {
+            io.passkey = passkey;
+            break;
+        }
+        case MP_BLUETOOTH_PASSKEY_ACTION_NUMCMP: {
```
`MP_BLUETOOTH_PASSKEY_ACTION_DISP` above and `MP_BLUETOOTH_PASSKEY_ACTION_NUMCMP` don't match the defines

jimmo commented:
> Minor note: in top post `MICROPY_BLUETOOTH_ENABLE_PAIRING_BONDING` should be `MICROPY_PY_BLUETOOTH_ENABLE_PAIRING_BONDING`

Fixed. (And the PASSKEY defines)

andrewleech commented:
My testing of this PR today has gone very well, though I think the IRK issue can cause some strange behaviors if there are multiple devices running. 

I've been testing primarily android central connecting to micropython peripheral on both unix (zephyr hci radio) and stm32 (mynewt/nimble hci radio).

Pairing and bonding works from both explicit pairing request and/or reading a characteristic with the `_FLAG_READ_ENCRYPTED` attribute.

This is with both the `ble_bonding_peripheral.py` example which uses a pin to pair (recognized as authenticated) and my custom application which has no pin (recognized as not authenticated). The link is encrypted and bonded in both cases.

The bond is maintained reliably across power cycles of either end of the link.

If the ble is disabled on micropython: `ble.active(False)` then enabled again `ble.active(True)` the bond doesn't seem to be persisted correctly, however I expect this is a pretty uncommon use case at this stage.

jimmo committed:
extmod/nimble: Generate and persist a unique IRK.

This provides a workaround for https://github.com/apache/mynewt-nimble/issues/887.

Without this, all devices would share a fixed default IRK.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
I've added a commit that adds a workaround for the fixed IRK.

andrewleech commented:
I've continued testing, still with micropython running as peripheral but a different Android (samsung A20, Android 10) phone running central (both nrf-connect and custom app).

With this phone, the pairing is not persisting over the micropython reboot when running on my custom stm32 application.
I can connect/disconnect/connect the phone while the stm32 is running just fine (pairing works) but after rebooting stm32 and re-connecting the pairing does not re-establish.

This exact same procedure with my previous Android phone (Xiaomi MI6 running Lineage OS Android 10) works correctly - pairing/bonding is maintained over stm32 reboots.

I'll continue to investigate the difference between the two.

dpgeorge commented:
> I can connect/disconnect/connect the phone while the stm32 is running just fine (pairing works) but after rebooting stm32 and re-connecting the pairing does not re-establish.

I'm seeing similar issues with Andriod-PYBD bonding.  Can you give a trace of the set/get secrets IRQ calls, the data (key and value) that each event is given?  (You may want to email me / obfuscate your public BD ADDR)

andrewleech commented:
Yeah the secret it's trying to read upon connection is the same each time connect/disconnect/re-connect without rebooting the mpy, but then it's different after reboot mpy. 

I suspect the peer IRK isn't working / not getting re-loaded into the radio correctly. I don't think the problem is from the latest "own IRK" commit though, I was seeing the same behavior before that too. I'd say my xiaomi is likely using a static public address so resolves consistently, whereas the samsung is probably using a randomised addr so relies on the IRK. 

Tomorrow I'll trace it all through before and after the reboot to confirm the behavior and hopefully find the fix.

dpgeorge commented:
> I'd say my xiaomi is likely using a static public address so resolves consistently, whereas the samsung is probably using a randomised addr so relies on the IRK.

Yes that could be it, sounds similar to what I'm seeing.

andrewleech commented:
Yep just confirmed; when the phones connect to ble_bonding_peripheral.py the working Xiaomi triggers a `ble_store_read_our_sec()` with `key_sec->peer_addr.type == BLE_ADDR_PUBLIC` while the samsung has the same function with `key_sec->peer_addr.type == BLE_ADDR_RANDOM`

andrewleech commented on examples/bluetooth/ble_bonding_peripheral.py:
```diff
@@ -0,0 +1,193 @@
+            if key is None:
+                for (t, _key), value in self._secrets.items():
+                    if t == sec_type:
+                        if i == index:
+                            return value
```
```suggestion
                        if i == index:
                            return value
                        i++                            
```

andrewleech replied:
@jimmo This fixes the "samsung not working over restart" issue I was having, turns out it's got nothing to do with the public/random address thing

dpgeorge replied:
yeah, I saw this as well

dpgeorge replied:
this is now fixed in #6662

dpgeorge commented:
@andrewleech does your BLE controller support LL privacy?

andrewleech commented:
> @andrewleech does your BLE controller support LL privacy?
I think so, I'm using mynewt/nimble hci controller build on stm32, zephyr hci on unix. Would need to double check the build configs to be sure though.

dpgeorge commented:
> Write multi-test.

Two basic multitests for pairing and bonding have been added in #6662. 

> PYBD doesn't work with RPA mode.

The BT controller on PYBD doesn't support LL privacy in the controller so RPA would need to be added on the host side.

dpgeorge commented:
I've tested this and am happy with how it works.

Merged in f2a9a0ac413cc337ccd78bbb26cd67cee7b1e210 through d79b9c6c7cb42c69c15d1b7869a04e889d812b82

Outstanding items can be addressed in follow-up work.

Thank you @jimmo and @andrewleech !
